File: sdklibrary/src/main/java/com/loner/android/sdk/webservice/network/networking/ServiceManager.java
package com.loner.android.sdk.webservice.network.networking;

import com.loner.android.sdk.utilities.Constant;
import com.loner.android.sdk.webservice.interfaces.APIsInterface;
import com.loner.android.sdk.webservice.interfaces.ActivityCallBackInterface;
import com.loner.android.sdk.webservice.interfaces.HTTPClientInterface;
import com.loner.android.sdk.webservice.models.BaseData;
import com.loner.android.sdk.webservice.network.apis.BaseRequest;
import com.loner.android.sdk.webservice.network.apis.RegisterRequest;
import com.loner.android.sdk.webservice.network.helper.NetworkErrorInformation;
import com.loner.android.sdk.webservice.network.helper.NetworkSuccessInformation;

import java.util.HashMap;

/**
 * This files used as ServiceManager. This class is responsible to create instances
 * of different API classes and call their respective Request functions.
 * Note : This is a Singleton class.
 */
public class ServiceManager implements HTTPClientInterface, APIsInterface {

    private static ServiceManager serviceManagerInstance = null;
    private RegisterRequest registerRequest;

    private HashMap<BaseRequest, ActivityCallBackInterface> callbackMap;

    private ServiceManager() {
        callbackMap = new HashMap<BaseRequest, ActivityCallBackInterface>();
    }

    /**
     * This method returns the ServiceManager instance.
     */
    public static ServiceManager getInstance() {
        if (serviceManagerInstance == null) {
            serviceManagerInstance = new ServiceManager();
        }
        return serviceManagerInstance;
    }

    @Override
    public void onSuccess(BaseRequest responseObject, BaseData baseDataObject, NetworkSuccessInformation networkSuccessInformation) {
        ActivityCallBackInterface listener = callbackMap.get(responseObject);
        if (listener != null) {
            listener.onResponseDataSuccess(networkSuccessInformation.getResponseMsg());
            callbackMap.remove(responseObject);
        }
    }

    @Override
    public void onFailure(BaseRequest responseObject, NetworkErrorInformation errorInformation) {
        ActivityCallBackInterface listener = callbackMap.get(responseObject);
        if (listener != null) {
            listener.onResponseDataFailure(errorInformation.getError());
            callbackMap.remove(responseObject);
        }
    }

    @Override
    public void registerDevice(String accessToken, ActivityCallBackInterface responseDataListener) {
        if(!accessToken.isEmpty()) {
            registerRequest = new RegisterRequest(accessToken, this);
            callbackMap.put(registerRequest, responseDataListener);
            registerRequest.register(registerRequest, Constant.STASK_Authentication);
        }else {
         responseDataListener.onResponseDataFailure("AccessToken can not empty");
        }
    }
}

File: sdklibrary/src/main/java/com/loner/android/sdk/webservice/network/networking/ServiceManager.kt
package com.loner.android.sdk.webservice.network.networking

import com.loner.android.sdk.utils.Constant
import com.loner.android.sdk.webservice.interfaces.APIsInterface
import com.loner.android.sdk.webservice.interfaces.ActivityCallBackInterface
import com.loner.android.sdk.webservice.interfaces.HTTPClientInterface
import com.loner.android.sdk.webservice.network.apis.AlertRequest
import com.loner.android.sdk.webservice.network.apis.BaseRequest
import com.loner.android.sdk.webservice.network.apis.RegisterRequest
import com.loner.android.sdk.webservice.network.helper.NetworkErrorInformation
import com.loner.android.sdk.webservice.network.helper.NetworkSuccessInformation
import java.util.*

/**
 * This files used as ServiceManager. This class is responsible to create instances
 * of different API classes and call their respective Request functions.
 * Note : This is a Singleton class.
 */
class ServiceManager private constructor() : HTTPClientInterface, APIsInterface {
    private var registerRequest: RegisterRequest? = null
    private var alertRequest: AlertRequest? = null
    private val callbackMap: HashMap<BaseRequest, ActivityCallBackInterface> = HashMap()

    override fun onSuccess(responseObject: BaseRequest, baseDataObject: Any, networkSuccessInformation: NetworkSuccessInformation) {
        val listener = callbackMap[responseObject]
        if (listener != null) {
            listener.onResponseDataSuccess(networkSuccessInformation.responseMsg!!)
            callbackMap.remove(responseObject)
        }
    }

    override fun onFailure(responseObject: BaseRequest, errorInformation: NetworkErrorInformation) {
        val listener = callbackMap[responseObject]
        if (listener != null) {
            listener.onResponseDataFailure(errorInformation.detailMessage!!)
            callbackMap.remove(responseObject)
        }
    }

    override fun registerDevice(accessToken: String, responseDataListener: ActivityCallBackInterface) =
            if (!accessToken.isEmpty()) {
                registerRequest = RegisterRequest(accessToken, this)
                callbackMap[registerRequest as BaseRequest] = responseDataListener
                registerRequest!!.register(registerRequest!!, Constant.STASK_Authentication)
            } else {
                responseDataListener.onResponseDataFailure("AccessToken can not empty")
            }


    override fun sendAlertApi(responseDataListener: ActivityCallBackInterface) {
            alertRequest = AlertRequest(this)
            callbackMap[alertRequest as BaseRequest] = responseDataListener
            alertRequest!!.alert(alertRequest!!, Constant.STASK_AlertEmergency)

    }


    companion object {
        @Volatile
        private var serviceManagerInstance: ServiceManager? = null
        fun getInstance(): ServiceManager {
            return serviceManagerInstance ?: synchronized(this) {
                ServiceManager().also {
                    serviceManagerInstance = it
                }
            }
        }
    }
}